File: src/effects/rainbow.rs
use std::time::Duration;
use crate::effects::{Effect, Buffer};

pub struct RainbowEffect {
    hue: f32,
}

impl RainbowEffect {
    pub fn new() -> Self {
        Self { hue: 0.0 }
    }
}

fn colorwheel(pos: u8) -> [u8; 3] {
    let pos = 255 - pos;
    if pos < 85 {
        [255 - pos * 3, 0, pos * 3]
    } else if pos < 170 {
        let pos = pos - 85;
        [0, pos * 3, 255 - pos * 3]
    } else {
        let pos = pos - 170;
        [pos * 3, 255 - pos * 3, 0]
    }
}

impl Effect for RainbowEffect {
    fn name(&self) -> &'static str { "Rainbow" }

    fn update(&mut self, buffer: &mut Buffer, delta: Duration) -> bool {
        self.hue = (self.hue + delta.as_secs_f32() * 50.0) % 256.0;
        let len = buffer.len();
        for (i, pixel) in buffer.iter_mut().enumerate() {
            *pixel = colorwheel(((self.hue as usize + i * 256 / len) % 256) as u8);
        }
        true
    }
}

File: effects/rainbow.lua
name = "Rainbow"

local hue = 0.0

local function colorwheel(pos)
    pos = (255 - pos) % 256
    if pos < 85 then
        return 255 - pos * 3, 0, pos * 3
    elseif pos < 170 then
        pos = pos - 85
        return 0, pos * 3, 255 - pos * 3
    else
        pos = pos - 170
        return pos * 3, 255 - pos * 3, 0
    end
end

function update(buf, dt)
    local n = buf:len()
    hue = (hue + dt * 50.0) % 256.0
    for i = 0, n - 1 do
        local r, g, b = colorwheel(math.floor(hue + i * 256 / n) % 256)
        buf:set(i, r, g, b)
    end
    return true
end
